File: ruby/lib/scheduler.rb
class ZipScheduler
  attr_reader :unfulfilled_orders

  def initialize(
    hospitals,
    num_zips,
    max_packages_per_zip,
    zip_speed_mps,
    max_cumulative_m
  )
    @hospitals = hospitals
    @num_zips = num_zips
    @max_packages_per_zip = max_packages_per_zip
    @zip_speed_mps = zip_speed_mps
    @zip_max_cumulative_range_m = max_cumulative_m

    # Track which orders haven't been launched yet
    @unfulfilled_orders = []
  end

  ##
  # Add a new order to our queue.
  #
  # Note: Called every time a new order arrives.
  #
  # @param [Order] order The order just placed.
  def queue_order(order)
    @unfulfilled_orders.append(order)

    # TODO: implement me
  end

  ##
  # Determines which flights should be launched right now. Each flight has an ordered list of
  # Orders to serve.
  #
  # Note: Will be called periodically (approximately once a minute).
  #
  # @param [Integer] current_time Seconds since midnight.
  # @return [Array] Flight objects that launch at this time.
  def launch_flights(current_time)
    # TODO: implement me!
    # You should remove any orders from `self.unfulfilled_orders` as you go
    []
  end
end

File: julia/scheduler.jl
# Everything in this file should be edited as needed
Base.@kwdef mutable struct ZipScheduler
    "Dictionary of hospital names to `Hospital` instances"
    hospital_directory::HospitalDirectory = HospitalDirectory()

    "Queue of unfulfilled orders"
    unfulfilled_orders::Vector{Order} = Order[]

    "Each Nest has this many Zips"
    num_zips::Int = 10

    """
    Each Zip can carry between 1 and this many packages per flight
    Note: a Zip can deliver more than 1 package per stop
    """
    max_packages_per_zip::Int = 3

    "Zips fly a constant groundspeed (m/s)"
    zip_speed_mps::Int = 30

    "Zips can fly a total roundtrip distance (m)"
    zip_max_cumulative_range_m::Int = 160 * 1000
end

"""
    queue_order!(scheduler::ZipScheduler, order::Order)

Add a new order to the queue
"""
function queue_order!(scheduler::ZipScheduler, order::Order)
    push!(scheduler.unfulfilled_orders, order)

    # TODO: Implement me!
    return nothing
end

"""
    launch_flights!(scheduler::ZipScheduler, current_time::Int)

Return a list of which flights should be launched right now
"""
function launch_flights!(scheduler::ZipScheduler, current_time::Int)
    # TODO: Implement me!
    # NOTE: Remove orders from `scheduler.unfulfilled_orders` as you go
    return Flight[]
end
